src/backend/distributed/executor/executor.c

```diff
@@ -499,7 +505,29 @@ StartDistributedExecution(DistributedExecution *execution)
 
 			if (TaskListRequires2PC(taskList))
 			{
+				/*
+				 * Although using two phase commit protocol is an independent decision than
+				 * failing on any error, we prefer to couple them. Our motivation is that
+				 * the failures are rare, and we prefer to avoid marking placements invalid
+				 * in case of failures.
+				 */
 				CoordinatedTransactionUse2PC();
+
+				execution->errorOnAnyFailure = true;
+			}
+			else if (MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC &&
+					 list_length(taskList) > 1)
+			{
+				Task *initialTask = (Task *) linitial(taskList);
+
+				if (ModifyTask(initialTask->taskType))
```

Comment:
Might be cleaner to use `IsModifyDistributedPlan`

Reply:
That function is probably not quite accurate because it does not consider DDL. Anyway, I'm guilty of using `initialTask` as well, but it's a bit of an anti-pattern.